unity/unity_utils.c

```diff
@@ -198,3 +200,216 @@ unity_mono_method_is_generic (MonoMethod* method)
+	char *dest = (char*)output;
+	int i;
+	guint32 num_named;
+	
+	MonoMethodSignature *sig = mono_method_signature(entry->ctor);
+	
+	// Skip prolog
+	src += 2;
+	for(i = 0; i < sig->param_count; ++i)
+	{
+		MonoType* paramType = sig->params[i];
+		int type = paramType->type;
+		
+		if (type == MONO_TYPE_VALUETYPE && paramType->data.klass->enumtype)
+			type = mono_class_enum_basetype(paramType->data.klass)->type;
+		
+		src += unpack_custom_attr_value(src, type, dest, image);
+		
+		// Advance dest as well
+		switch(type)
+		{
+			case MONO_TYPE_U1:
+			case MONO_TYPE_I1:
+			case MONO_TYPE_BOOLEAN:
+				dest += 1;
+				break;
+			case MONO_TYPE_CHAR:
+			case MONO_TYPE_U2:
+			case MONO_TYPE_I2:
+				dest += 2;
+				break;
+#if SIZEOF_VOID_P == 4
+			case MONO_TYPE_U:
+			case MONO_TYPE_I:
+#endif
+			case MONO_TYPE_R4:
+			case MONO_TYPE_U4:
+			case MONO_TYPE_I4:
+				dest += 4;
+				break;
+#if SIZEOF_VOID_P == 8
+			case MONO_TYPE_U:
+			case MONO_TYPE_I:
+#endif
+			case MONO_TYPE_U8:
+			case MONO_TYPE_I8:
+			case MONO_TYPE_R8:
+				dest += 8;
+				break;
+			case MONO_TYPE_STRING:
+				dest += sizeof(UnityMonoMetadataString*);
+				break;
+			case MONO_TYPE_CLASS:
+				dest += sizeof(MonoType*);
+				break;
+			default:
+				g_error("Unknown/unsupported parameter type");
+				break;
+		}
+	}
+	
+	// Now, named parameters
+	if(namedParameterHandler == NULL) return;
+	num_named = read16(src);
+	src += 2;
+	for (i = 0; i < num_named; ++i)
+	{
+		UnityMonoMetadataString* paramName;
+		int type;
+		// The value will be unpacked into this buffer. The actual value size depends on the type,
+		// but the largest thing we can unpack right now is an int64, so an 8 byte buffer should be sufficient.
+		char buffer[8];
+		
+		src++; // skip the field/property specifier
+		type = *(src++);
+		
+		if (type == MONO_TYPE_SZARRAY)
+		{
+			g_error("Array types are not supported for named parameter unpacking");
+			type = *(src++);
+		}
+		
+		if (type == MONO_TYPE_ENUM)
+		{
+			// Enum typed parameters pack the actual name of the enum type
+			// We don't really care about it for our purposes, just skip past it
+			gint type_len = mono_metadata_decode_blob_size (src, &src);
+			src += type_len;
+		}
+		
+		paramName = (UnityMonoMetadataString*)src;
+		src += mono_metadata_decode_blob_size (src, &src);
+		
+		src += unpack_custom_attr_value(src, type, buffer, image);
+		
+		namedParameterHandler(output, paramName, type, &buffer[0]);
+	}
+}
+
+guint32 unity_mono_metadata_string_length(UnityMonoMetadataString* string)
+{
+	if(string == NULL) return 0;
+	return mono_metadata_decode_blob_size(string, &string);
+}
+
+void unity_mono_metadata_string_copy(UnityMonoMetadataString* string, char* buffer, size_t bufferSize)
+{
+	guint32 strLen;
+	if(bufferSize == 0) return;
+	if(string == NULL)
+	{
+		buffer[0] = 0;
+		return;
+	}
+	strLen = mono_metadata_decode_blob_size(string, &string);
+	if(strLen + 1 > bufferSize)
+		strLen = bufferSize - 1;
```

Comment:
I think it would be nice if this function returned the number of bytes that did not fit into the buffer. I realize that unity_mono_metadata_string_length provides the required size of the buffer, but I still think it would be nice.